Mathlib/Topology/Algebra/InfiniteSum/NatInt.lean

```diff
@@ -321,17 +259,236 @@ end UniformGroup
+      forall_apply_eq_imp_iff₂, not_false_eq_true, implies_true, forall_const]
+
+lemma Summable.natCast_add_negSucc {f : ℤ → α} (hf : Summable f) :
+    Summable (fun n : ℕ ↦ f n + f (Int.negSucc n)) :=
+  hf.hasSum.natCast_add_negSucc.summable
+
+lemma tsum_natCast_add_negSucc [T2Space α] {f : ℤ → α} (hf : Summable f) :
+    ∑' (n : ℕ), (f n + f (Int.negSucc n)) = ∑' (n : ℤ), f n :=
+  hf.hasSum.natCast_add_negSucc.tsum_eq
+
+section ContinuousAdd
+
+variable [ContinuousAdd α]
+
+lemma HasSum.of_natCast_negSucc {f : ℤ → α}
+    (hf₁ : HasSum (fun n : ℕ ↦ f n) a) (hf₂ : HasSum (fun n : ℕ ↦ f (Int.negSucc n)) b) :
+    HasSum f (a + b) := by
+  have hi₁ : Injective ((↑) : ℕ → ℤ) := @Int.ofNat.inj
+  have hi₂ : Injective Int.negSucc := @Int.negSucc.inj
+  have : IsCompl (Set.range ((↑) : ℕ → ℤ)) (Set.range Int.negSucc) := by
+    constructor
+    · rw [disjoint_iff_inf_le]
+      rintro _ ⟨⟨i, rfl⟩, ⟨j, ⟨⟩⟩⟩
+    · rw [codisjoint_iff_le_sup]
+      rintro (i | j) <;> simp
+  exact (hi₁.hasSum_range_iff.mpr hf₁).add_isCompl this (hi₂.hasSum_range_iff.mpr hf₂)
+#align has_sum.nonneg_add_neg HasSum.of_natCast_negSucc
+
+-- deprecated 2024-03-04
+@[deprecated] alias HasSum.nonneg_add_neg := HasSum.of_natCast_negSucc
+
+lemma Summable.of_natCast_negSucc {f : ℤ → α}
+    (hf₁ : Summable fun n : ℕ ↦ f n)  (hf₂ : Summable fun n : ℕ ↦ f (Int.negSucc n)) :
+    Summable f :=
+  (hf₁.hasSum.of_natCast_negSucc hf₂.hasSum).summable
+
+lemma tsum_of_natCast_negSucc [T2Space α] {f : ℤ → α}
+    (hf₁ : Summable fun n : ℕ ↦ f n) (hf₂ : Summable fun n : ℕ ↦ f (Int.negSucc n)) :
+    ∑' n : ℤ, f n = ∑' n : ℕ, f n + ∑' n : ℕ, f (Int.negSucc n) :=
+  (hf₁.hasSum.of_natCast_negSucc hf₂.hasSum).tsum_eq
+
+/-- If `f₀, f₁, f₂, ...` and `g₀, g₁, g₂, ...` have sums `a`, `b` respectively, then the `ℤ`-indexed
+sequence: `..., g₂, g₁, g₀, f₀, f₁, f₂, ...` (with `f₀` at the `0`-th position) has sum `a + b`. -/
+lemma HasSum.int_rec {f g : ℕ → α} (hf : HasSum f a) (hg : HasSum g b) :
+    HasSum (Int.rec f g) (a + b) :=
+  HasSum.of_natCast_negSucc hf hg
+#align has_sum.int_rec HasSum.int_rec
+
+/-- If `f₀, f₁, f₂, ...` and `g₀, g₁, g₂, ...` are both summable then so is the `ℤ`-indexed
+sequence: `..., g₂, g₁, g₀, f₀, f₁, f₂, ...` (with `f₀` at the `0`-th position). -/
+lemma Summable.int_rec {f g : ℕ → α} (hf : Summable f) (hg : Summable g) : Summable (Int.rec f g) :=
+  .of_natCast_negSucc hf hg
+
+/-- If `f₀, f₁, f₂, ...` and `g₀, g₁, g₂, ...` are both summable, then the sum of the `ℤ`-indexed
+sequence: `..., g₂, g₁, g₀, f₀, f₁, f₂, ...` (with `f₀` at the `0`-th position) is
+`∑' n, f n + ∑' n, g n`. -/
+lemma tsum_int_rec [T2Space α] {f g : ℕ → α} (hf : Summable f) (hg : Summable g) :
+    ∑' n : ℤ, Int.rec f g n = ∑' n : ℕ, f n + ∑' n : ℕ, g n :=
+  (hf.hasSum.int_rec hg.hasSum).tsum_eq
+
+theorem HasSum.natCast_add_neg_natCast {f : ℤ → α} (hf : HasSum f a) :
+    HasSum (fun n : ℕ ↦ f n + f (-n)) (a + f 0) := by
```

Comment:
I thought that the proof of this should be much easier, but then I realized that the results I wanted to use have stronger assumptions (`AddCommGroup` instead of `AddCommMonoid`, `T2Space`, ...).
Maybe it's worth adding a comment (or docstring) mentioning that this uses only fairly week assumptions on `α`?

Reply:
Funnily enough, I originally introduced this lemma myself in a mathlib3 PR over a year ago, and what I wrote was a lemma about topological groups with a proof about 5 lines long. The reviewer (Sebastian Gouëzel) generalized my simple little lemma into the behemoth you see here. (The only change to this lemma in the present PR is renaming it & moving it around slightly.)